### PR TITLE
fix(bot): expand SoundCloud short links before discord-player resolution

### LINKED ISSUE #1162: bug(bot): expand SoundCloud short URLs before passing to discord-player resolver
## What
SoundCloud short links (`on.soundcloud.com/...`) are passed directly to discord-player's `attachmentextractor`, which can't resolve them. The user gets a `NoResultError` instead of a helpful message.

## Evidence
Sentry LUCKY-3R (2 events) — `NoResultError` from `com.discord-player.attachmentextractor` in `resolveQueryWithFallbacks`

## Acceptance criteria
- [ ] `on.soundcloud.com` short URLs are detected and expanded (HTTP follow redirect) before passing to discord-player
- [ ] LUCKY-3R stops firing in Sentry
- [ ] `/play https://on.soundcloud.com/...` resolves and queues correctly

<sub>Surfaced by `/backlog` on 2026-06-04. Dedup key: `audit-deep:bug:soundcloud-shortlink`</sub>

### PR DESCRIPTION
**Pull request by LucasSantana-Dev**
Closes #1162 · Resolves Sentry LUCKY-3R

`on.soundcloud.com` short links were passed straight to discord-player, which can't resolve them → `NoResultError`. Now expanded via HEAD redirect-follow before resolution.

**Security:** only `on.soundcloud.com` inputs are expanded; the resolved host must be `soundcloud.com` or `*.soundcloud.com` (exact/suffix match — rejects `soundcloud.com.attacker.com`). 5s timeout; graceful fallback to original URL on any failure.

- [ ] 2500 bot tests green (incl. happy path, malicious-redirect, domain-suffix bypass, timeout)
- [ ] tsc clean

---
## Summary by cubic
Fixes failures when users paste SoundCloud short links by expanding `on.soundcloud.com` URLs before resolution. Prevents `NoResultError` in `discord-player` and blocks unsafe redirects.

- **Bug Fixes**
  - Expand only `on.soundcloud.com` via HEAD + redirect follow with `withTimeout` (5s), then normalize.
  - Validate resolved host equals `soundcloud.com` or ends with `.soundcloud.com`; reject suffix lookalikes.
  - Fall back to the original URL on timeout, network errors, or invalid/malicious redirects.
  - Tests cover happy path, non-short-link, malformed URL, malicious redirect, and domain-suffix bypass.

<sup>Written for commit cc74a7d08ee8cce1eee9db90de024acb30e36901. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/LucasSantana-Dev/Lucky/pull/1177?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## New Features
* Added support for SoundCloud short links in the music player, automatically expanding them to full URLs for improved usability.

## Tests
* Added comprehensive test coverage for SoundCloud short link expansion and URL normalization, including edge cases and security validation.

### CODE REVIEW

**Comment by LucasSantana-Dev** [comment]
@cubic-dev-ai review

**Comment by LucasSantana-Dev** [comment]
@cubic-dev-ai review

File: packages/bot/src/functions/music/commands/play/queryUtils.ts [hunk 1]
```diff
@@ -31,6 +32,82 @@
...
+ * Expands SoundCloud short links (on.soundcloud.com) by following the HTTP redirect.
+ * Returns the expanded URL if successful, or the original URL if expansion fails.
+ * Uses a 5-second timeout to prevent hanging requests.
+ *
+ * Security: only expands on.soundcloud.com hosts and validates the resolved URL
+ * is a soundcloud.com domain before returning it.
+ */
+export async function expandSoundCloudShortUrl(url: string): Promise<string> {
+    // Fast path: not a short link
+    if (!url.includes('on.soundcloud.com')) {
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/queryUtils.ts:45]: Addressed in the latest commit: host matched via URL parse + exact lowercase hostname (fixes substring/case), fetch aborted on timeout via AbortController, and contribution-limit check moved before expansion. +tests.

File: packages/bot/src/functions/music/commands/play/handlers/playHandler.ts [hunk 2]
```diff
@@ -55,7 +56,10 @@ export async function executePlayHandler({
     }
 
     const rawQuery = interaction.options.getString('query', true)
-    const query = normalizeSoundCloudUrl(rawQuery)
+    // Expand SoundCloud short links first (on.soundcloud.com → full URL)
+    const expandedQuery = await expandSoundCloudShortUrl(rawQuery)
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/handlers/playHandler.ts:60]: Addressed in the latest commit: host matched via URL parse + exact lowercase hostname (fixes substring/case), fetch aborted on timeout via AbortController, and contribution-limit check moved before expansion. +tests.

File: packages/bot/src/functions/music/commands/play/queryUtils.ts [hunk 3]
```diff
@@ -31,6 +32,82 @@ export function isUrl(query: string): boolean {
...
+        }
+
+        // Follow redirects with a 5-second timeout, using HEAD first (no body download)
+        const expanded = await withTimeout(
+            (async () => {
+                try {
+                    const response = await (global.fetch as typeof fetch)(url, {
+                        method: 'HEAD',
+                        redirect: 'follow',
+                    })
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/queryUtils.ts:62]: Addressed in the latest commit: host matched via URL parse + exact lowercase hostname (fixes substring/case), fetch aborted on timeout via AbortController, and contribution-limit check moved before expansion. +tests.

File: packages/bot/src/functions/music/commands/play/queryUtils.ts [hunk 4]
```diff
@@ -31,6 +32,82 @@ export function isUrl(query: string): boolean {
...
+    // Fast path: not a short link
+    if (!url.includes('on.soundcloud.com')) {
+        return url
+    }
+
+    try {
+        const parsed = new URL(url)
+        if (parsed.hostname !== 'on.soundcloud.com') {
+            return url
+        }
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/queryUtils.ts:53]: Addressed in the latest commit: host matched via URL parse + exact lowercase hostname (fixes substring/case), fetch aborted on timeout via AbortController, and contribution-limit check moved before expansion. +tests.

File: packages/bot/src/functions/music/commands/play/queryUtils.ts [hunk 5]
```diff
@@ -31,6 +32,82 @@ export function isUrl(query: string): boolean {
...
+
+        // Follow redirects with a 5-second timeout, using HEAD first (no body download)
+        const expanded = await withTimeout(
+            (async () => {
+                try {
+                    const response = await (global.fetch as typeof fetch)(url, {
+                        method: 'HEAD',
+                        redirect: 'follow',
+                    })
+                    const finalUrl = response.url
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/queryUtils.ts:63]: Not exploitable in this context: the fetch is only ever issued to a URL whose hostname is exactly on.soundcloud.com (validated before the request), uses HEAD (no body), and the final response.url hostname is validated to be soundcloud.com / *.soundcloud.com. The redirect chain is controlled by SoundCloud's own short-link service — an external user supplying a play URL cannot make on.soundcloud.com redirect to internal infra, so there is no attacker-controlled SSRF path. Timed-out fetches are now aborted via AbortController. Resolving as a contextual false-positive.

File: packages/bot/src/functions/music/commands/play/queryUtils.ts [hunk 6]
```diff
@@ -31,6 +32,82 @@ export function isUrl(query: string): boolean {
...
+                    if (
+                        finalParsed.hostname !== 'soundcloud.com' &&
+                        !finalParsed.hostname?.endsWith('.soundcloud.com')
+                    ) {
+                        // Redirect went somewhere unexpected — reject and fall back
+                        throw new Error(
+                            `Redirect destination is not a soundcloud.com domain: ${finalUrl}`,
+                        )
+                    }
+
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/functions/music/commands/play/queryUtils.ts:76]: Not exploitable in this context: the fetch is only ever issued to a URL whose hostname is exactly on.soundcloud.com (validated before the request), uses HEAD (no body), and the final response.url hostname is validated to be soundcloud.com / *.soundcloud.com. The redirect chain is controlled by SoundCloud's own short-link service — an external user supplying a play URL cannot make on.soundcloud.com redirect to internal infra, so there is no attacker-controlled SSRF path. Timed-out fetches are now aborted via AbortController. Resolving as a contextual false-positive.
